internal/scheduler/scheduler.go

```diff
@@ -0,0 +1,181 @@
+// Package scheduler implements a simple scheduler that triggers the next item when its due time is
+// reached based on the list of upcoming items.
+package scheduler
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/kopia/kopia/internal/clock"
+	"github.com/kopia/kopia/internal/ctxutil"
+	"github.com/kopia/kopia/repo/logging"
+)
+
+var log = logging.Module("scheduler")
+
+// GetItemsFunc is a callback that returns items for the scheduler to consider.
+type GetItemsFunc func(ctx context.Context, now time.Time) []Item
+
+// Item is a describes item that can be scheduled with a function invoked the next time the item is due.
+type Item struct {
+	Description string
+	NextTime    time.Time
+	Trigger     func()
+}
+
+// Scheduler is a scheduler triggers the next upcoming events based on the list of items returned by listItems().
+type Scheduler struct {
+	TimeNow func() time.Time
+	Debug   bool
+
+	refreshRequested chan string
+	getItems         GetItemsFunc
+	closed           chan struct{}
+	wg               sync.WaitGroup
+}
+
+// Options provides options for the scheduler.
+type Options struct {
+	TimeNow        func() time.Time
+	Debug          bool
+	RefreshChannel chan string
+}
+
+// Start runs a new scheduler that will call getItems() to get the list of items to schedule.
+func Start(ctx context.Context, getItems GetItemsFunc, opts Options) *Scheduler {
+	timeNow := opts.TimeNow
+
+	if timeNow == nil {
+		timeNow = clock.Now
+	}
+
+	s := &Scheduler{
+		TimeNow:          timeNow,
+		refreshRequested: opts.RefreshChannel,
+		closed:           make(chan struct{}),
+		getItems:         getItems,
+		Debug:            opts.Debug,
+	}
+
+	s.wg.Add(1)
+
+	go func() {
+		defer s.wg.Done()
+
+		s.run(ctxutil.Detach(ctx))
+	}()
+
+	return s
+}
+
+const sleepTimeWhenNoUpcomingSnapshots = 8 * time.Hour
+
+func (s *Scheduler) upcomingItems(ctx context.Context, now time.Time) (nextTriggerTime time.Time, toTrigger []Item) {
+	allsm := s.getItems(ctx, now)
+
+	for _, t := range allsm {
+		if nextTriggerTime.IsZero() || t.NextTime.Before(nextTriggerTime) {
+			nextTriggerTime = t.NextTime
+			toTrigger = nil
```

Comment:
Can getItems() return items which were supposed to be triggered in past ?

So, even when we have found a new smaller trigger time here, don't we at-least want to keep those items which have a trigger time already in the past from now ? (i.e. t.NextTime < now)

Reply:
yes, there's a test for it.